### ISSUE TITLE
Incomplete S3 object created when source file is deleted during upload

### ISSUE THREAD

**Issue by sdcu** [rclone/rclone#7449]
#### What is the problem you are having with rclone?

An incomplete object is created at the destination S3 bucket when the source file is deleted during a multi-thread copy.
This incomplete object could overwrite an existing object at the destination bucket, causing data loss.

#### What is your rclone version (output from `rclone version`)

```
rclone v1.65.0-beta.7480.e1b0417c2
- os/version: debian 12.2 (64 bit)
- os/kernel: 5.4.0-26-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.21.4
- go/linking: static
- go/tags: none
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Debian GNU/Linux 12 (bookworm) amd64

#### Which cloud storage system are you using?

 AWS S3

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

To reproduce, delete the source file after the first multi-thread chunk copy starts:

```
rclone -vv copyto --s3-no-check-bucket --s3-disable-checksum test.bin  aws-s3:bucket/test/test.bin
```

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
2023/11/16 17:44:55 DEBUG : rclone: Version "v1.65.0-beta.7480.e1b0417c2" starting with parameters ["rclone" "-vv" "copyto" "--s3-no-check-bucket" "--s3-disable-checksum" "test.bin" "aws-s3:bucket/test/test.bin"]
2023/11/16 17:44:55 DEBUG : Creating backend with remote "test.bin"
2023/11/16 17:44:55 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
2023/11/16 17:44:55 DEBUG : fs cache: adding new entry for parent of "test.bin", "/"
2023/11/16 17:44:55 DEBUG : Creating backend with remote "aws-s3:bucket/test/"
2023/11/16 17:44:55 DEBUG : aws-s3: detected overridden config - adding "{YkiGG}" suffix to name
2023/11/16 17:44:55 DEBUG : fs cache: renaming cache item "aws-s3:bucket/test/" to be canonical "aws-s3{YkiGG}:bucket/test"
2023/11/16 17:44:55 DEBUG : test.bin: Need to transfer - File not found at Destination
2023/11/16 17:44:56 DEBUG : test.bin: open chunk writer: started multipart upload: VAgrIIHnaN.Ka1UDWNi32pxTNE62Yl0lbG1kd3DTfheyk6y9VF724Rpei9OGR6g7RiiS8y96qaQSlxLcpWH9sOb6jKINWY0RoodzTxtSqfOF7sW4pq8i0WwZOUUG_o0k
2023/11/16 17:44:56 DEBUG : test.bin: multi-thread copy: using backend concurrency of 4 instead of --multi-thread-streams 4
2023/11/16 17:44:56 DEBUG : test.bin: Starting multi-thread copy with 3532 chunks of size 5Mi with 4 parallel streams
2023/11/16 17:44:56 DEBUG : test.bin: multi-thread copy: chunk 4/3532 (15728640-20971520) size 5Mi starting
2023/11/16 17:44:56 DEBUG : test.bin: multi-thread copy: chunk 2/3532 (5242880-10485760) size 5Mi starting
2023/11/16 17:44:56 DEBUG : test.bin: multi-thread copy: chunk 1/3532 (0-5242880) size 5Mi starting
2023/11/16 17:44:56 DEBUG : test.bin: multi-thread copy: chunk 3/3532 (10485760-15728640) size 5Mi starting
2023/11/16 17:45:01 DEBUG : test.bin: multipart upload wrote chunk 4 with 5242880 bytes and etag "dcdb95fd7a4fdb51e4572779e5b1ecaf"
2023/11/16 17:45:01 DEBUG : test.bin: multi-thread copy: chunk 4/3532 (15728640-20971520) size 5Mi finished
2023/11/16 17:45:01 DEBUG : test.bin: multi-thread copy: chunk 5/3532 (20971520-26214400) size 5Mi starting
2023/11/16 17:45:02 DEBUG : test.bin: multipart upload wrote chunk 1 with 5242880 bytes and etag "4c2de5f1b5ba6ca618b547a355078e28"
2023/11/16 17:45:02 DEBUG : test.bin: multi-thread copy: chunk 1/3532 (0-5242880) size 5Mi finished
2023/11/16 17:45:02 DEBUG : test.bin: multi-thread copy: chunk 6/3532 (26214400-31457280) size 5Mi starting
2023/11/16 17:45:02 DEBUG : test.bin: multipart upload wrote chunk 2 with 5242880 bytes and etag "1f584c6343dd2171392558db4777ded1"
2023/11/16 17:45:02 DEBUG : test.bin: multi-thread copy: chunk 2/3532 (5242880-10485760) size 5Mi finished
2023/11/16 17:45:02 DEBUG : test.bin: multi-thread copy: chunk 7/3532 (31457280-36700160) size 5Mi starting
2023/11/16 17:45:05 DEBUG : test.bin: multipart upload wrote chunk 3 with 5242880 bytes and etag "00b5d0e6f6f4fd8579cb70aa8402a386"
2023/11/16 17:45:05 DEBUG : test.bin: multi-thread copy: chunk 3/3532 (10485760-15728640) size 5Mi finished
2023/11/16 17:45:05 DEBUG : test.bin: multi-thread copy: chunk 8/3532 (36700160-41943040) size 5Mi starting
2023/11/16 17:45:05 DEBUG : test.bin: multi-thread copy: chunk 8/3532 failed: multi-thread copy: failed to open source: lstat /test.bin: no such file or directory
2023/11/16 17:45:05 DEBUG : test.bin: multi-thread copy: chunk 9/3532 (41943040-47185920) size 5Mi starting
2023/11/16 17:45:05 DEBUG : test.bin: multi-thread copy: chunk 9/3532 failed: multi-thread copy: failed to open source: lstat /test.bin: no such file or directory
2023/11/16 17:45:05 DEBUG : test.bin: multi-thread copy: chunk 5/3532 failed: multi-thread copy: failed to write chunk: failed to upload chunk 5 with 5242880 bytes: RequestCanceled: request context canceled
caused by: context canceled
2023/11/16 17:45:05 DEBUG : test.bin: multi-thread copy: chunk 7/3532 failed: multi-thread copy: failed to write chunk: failed to upload chunk 7 with 5242880 bytes: RequestCanceled: request context canceled
caused by: context canceled
2023/11/16 17:45:05 DEBUG : test.bin: multi-thread copy: chunk 6/3532 failed: multi-thread copy: failed to write chunk: failed to upload chunk 6 with 5242880 bytes: RequestCanceled: request context canceled
caused by: context canceled
2023/11/16 17:45:06 DEBUG : test.bin: multipart upload "VAgrIIHnaN.Ka1UDWNi32pxTNE62Yl0lbG1kd3DTfheyk6y9VF724Rpei9OGR6g7RiiS8y96qaQSlxLcpWH9sOb6jKINWY0RoodzTxtSqfOF7sW4pq8i0WwZOUUG_o0k" finished
2023/11/16 17:45:06 DEBUG : test.bin: multi-thread copy: cancelling transfer on exit
2023/11/16 17:45:06 DEBUG : test.bin: multipart upload "VAgrIIHnaN.Ka1UDWNi32pxTNE62Yl0lbG1kd3DTfheyk6y9VF724Rpei9OGR6g7RiiS8y96qaQSlxLcpWH9sOb6jKINWY0RoodzTxtSqfOF7sW4pq8i0WwZOUUG_o0k" aborted
2023/11/16 17:45:06 ERROR : test.bin: Failed to copy: multi-thread copy: failed to open source: lstat /test.bin: no such file or directory
2023/11/16 17:45:06 ERROR : Attempt 1/3 failed with 1 errors and: multi-thread copy: failed to open source: lstat /test.bin: no such file or directory
2023/11/16 17:45:06 ERROR : Attempt 2/3 failed with 1 errors and: object not found
2023/11/16 17:45:06 ERROR : Attempt 3/3 failed with 1 errors and: object not found
2023/11/16 17:45:06 INFO  :
Transferred:           35 MiB / 35 MiB, 100%, 3.500 MiB/s, ETA 0s
Errors:                 1 (retrying may help)
Elapsed time:        11.2s

2023/11/16 17:45:06 DEBUG : 8 go routines active
2023/11/16 17:45:06 Failed to copyto: object not found
```
Bucket name, etags, and upload id are replaced with dummy values.

##### Ls before the copyto command

```
# rclone ls test.bin
18517852160 test.bin
```
```
# rclone lsjson aws-s3:bucket/test/test.bin
[
]
```

#### Ls after the copyto command

```
rclone ls aws-s3:bucket/test/test.bin
 20971520 test.bin
```

#### Possible Cause

The issue seems to lie in the multiThreadCopy function, which calls [chunkWriter.Close](https://github.com/rclone/rclone/blob/e1b0417c286bdcf5e866d70d51cb75be179c25e6/fs/operations/multithread.go#L219) before returning.

This results in the deferred closure calling [chunkWriter.Abort](https://github.com/rclone/rclone/blob/e1b0417c286bdcf5e866d70d51cb75be179c25e6/fs/operations/multithread.go#L164) after the S3 backend has already called [CompleteMultipartUploadWithContext](https://github.com/rclone/rclone/blob/e1b0417c286bdcf5e866d70d51cb75be179c25e6/backend/s3/s3.go#L5775), causing the file to be incomplete.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

<sub>Alexander Hunt <alexander.hunt@mercedes-benz.com>, Mercedes-Benz Tech Innovation GmbH,
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)</sub>

**Comment by ncw** [comment]
Here is my attempt at a fix - please give it a go!

It has an integration tests to make sure this doesn't happen again. This was quite complicated to write!

[v1.65.0-beta.7526.105b76fce.fix-7449-s3-multipart](https://beta.rclone.org/branch/fix-7449-s3-multipart/v1.65.0-beta.7526.105b76fce.fix-7449-s3-multipart/) on branch [fix-7449-s3-multipart](https://github.com/rclone/rclone/tree/fix-7449-s3-multipart) (uploaded in 15-30 mins)

**Comment by sdcu** [comment]
Thank you for your effort in quickly fixing the issue.

I noticed that the race test seems to have failed in the [GitHub action](https://github.com/rclone/rclone/actions/runs/6961072130).

However, after building from source and conducting manual testing, I'm happy to confirm that the fix works exactly as expected, when the source file is deleted during upload.

Thank you very much!

**Comment by ncw** [comment]
Thank you for testing.

Hmm, yes the tests are using a pathological amount of CPU. Hopefully this one will build and work

[v1.65.0-beta.7529.6e06f363b.fix-7449-s3-multipart](https://beta.rclone.org/branch/fix-7449-s3-multipart/v1.65.0-beta.7529.6e06f363b.fix-7449-s3-multipart/) on branch [fix-7449-s3-multipart](https://github.com/rclone/rclone/tree/fix-7449-s3-multipart) (uploaded in 15-30 mins)

If it does I'll merge it for v1.65

**Comment by sdcu** [comment]
I tested the newest beta version (v1.65.0-beta.7545.0244caf13) and could not reproduce the behavior anymore. 
The issue is resolved for us.

Thank you very much!
